// react-jsonschema-form-pagination/lib/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.findFieldNavs = findFieldNavs;
exports.getNavAliases = getNavAliases;
exports.isDevelopment = isDevelopment;
exports.toArray = toArray;
exports.isEmptySchema = isEmptySchema;
var GENERIC_NAV = exports.GENERIC_NAV = "default";
var UI_ORDER = exports.UI_ORDER = "ui:order";
var UI_NAV_ID = exports.UI_NAV_ID = "nav";

function findFieldNavs(field, uiSchema) {
  var navs = uiSchema[field] && uiSchema[field][UI_NAV_ID] ? uiSchema[field][UI_NAV_ID] : [];
  return toArray(navs);
}

function getNavAliases(_ref) {
  var _ref$navConf = _ref.navConf;
  _ref$navConf = _ref$navConf === undefined ? {} : _ref$navConf;
  var _ref$navConf$aliases = _ref$navConf.aliases,
      aliases = _ref$navConf$aliases === undefined ? {} : _ref$navConf$aliases;

  return aliases;
}

function isDevelopment() {
  return process.env.NODE_ENV !== "production";
}

var toError = exports.toError = function toError(message) {
  if (isDevelopment()) {
    throw new ReferenceError(message);
  } else {
    console.error(message);
  }
};

function toArray(val) {
  if (Array.isArray(val)) {
    return val;
  } else {
    return [val];
  }
}

function isEmptySchema(schema) {
  return !schema || !schema.properties || Object.keys(schema.properties).length === 0;
}
